src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -411,6 +438,9 @@ static inline int MPIDI_NM_mpi_put(const void *origin_addr,
     size_t target_bytes, origin_bytes;
     MPI_Aint origin_true_lb, target_true_lb;
     size_t offset;
+    struct fi_msg_rma msg;
+    struct iovec iov;
+    struct fi_rma_iov riov;
 
     if (!MPIDI_OFI_ENABLE_RMA) {
         mpi_errno = MPIDI_CH4U_mpi_put(origin_addr, origin_count, origin_datatype,
```

Comment:
I'd recommend that you move the entire issuing logic including PROC_NULL target, zero-byte checking, self-target, and contig-small message into `do_put`, thus the same logic can be reused in `rput`. The only difference of `rput` is that `rput` returns a request. Same for `get|rget`.

The login in `rput|rget` can be:
- If `do_put` or `do_get` returns a request, then just return it.
- If `do_put` or `do_get` does not return the request (e.g., target == PROC_NULL), then create a completed request here and return it.

Please also make sure that code refactoring and the segment fix are in two separate commits.